### DISCUSSION TITLE
> It is amazing that it's been 4 years.

### DISCUSSION THREAD

**Discussion by kiwavi** [prisma/prisma#25424] [Q&A] [answered]
> It is amazing that it's been 4 years.
> 
> `@default(dbgenerated("NOW()"))` seems work. However, `@updatedAt` also doesn't respect timezone neither. So it is useless.
> <img alt="Screenshot 2024-06-07 at 2 33 49 PM" width="450" src="https://private-user-images.githubusercontent.com/19797697/337535924-967116bd-16a5-4e37-a597-9f1972e30298.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjg3NTI2ODUsIm5iZiI6MTcyODc1MjM4NSwicGF0aCI6Ii8xOTc5NzY5Ny8zMzc1MzU5MjQtOTY3MTE2YmQtMTZhNS00ZTM3LWE1OTctOWYxOTcyZTMwMjk4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDEyVDE2NTk0NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWM4YWE5NjU0ZWZiMWIwNDcwNDI0YWVjY2IwNTA2OTU5MWFiYWY3OTIyOTkyODNhYzY3MDFmNThjMTM1MzkyYjQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.9ELcdBZSoJB2Fehqa5rAByYRIS-cDNjtxDU007vkzP8">
> 
> Besides, it's interesting that `dbgenerated("NOW()")` and `@updatedAt` are not even synced.

Interesting how this is handled under the hood. Because when I look at the migration file the difference is between current_timestamp and now. For dbgenerated, the command is `TIMESTAMP(3) NOT NULL DEFAULT NOW()` while for the datetime with default(now) it is `TIMESTAMP(3) NOT NULL DEFAULT CURRENT_TIMESTAMP`. In postgres these two commands basically do the same thing. So where could the difference be coming from?

What makes `TIMESTAMP(3) NOT NULL DEFAULT NOW()` respect timezone while `TIMESTAMP(3) NOT NULL DEFAULT CURRENT_TIMESTAMP` does not?

_Originally posted by @kiwavi in https://github.com/prisma/prisma/issues/5051#issuecomment-2408896801_

**Comment by ludralph** [comment, marked as answer]
Hi @kiwavi 

When you use `@default(now())` in Prisma, the timestamp is generated by Prisma’s Query Engine at the time the query is executed. This means that the current timestamp is assigned by the Prisma layer, and while it's reflected in the Prisma schema, it’s not directly tied to the database's internal timestamping mechanisms. See this [open issue](https://github.com/prisma/prisma/issues/1831)

In contrast, `@default(dbgenerated("NOW()"))` relies on the database’s native function (`NOW()` in PostgreSQL) to generate the timestamp. In this case, the database itself is responsible for assigning the current timestamp when the record is created, ensuring the timestamp is generated at the database level.

Although `NOW()` and `CURRENT_TIMESTAMP` are functionally equivalent in PostgreSQL (both return the current timestamp), the key difference lies in how and where they are applied:

`@default(now())`: Timestamp is assigned by Prisma’s Query Engine, which might introduce minor differences due to application-server clock variations or latency.
`@default(dbgenerated("NOW()"))`: Timestamp is assigned natively by the database, ensuring consistency for records created directly within the database, even outside Prisma.

This difference in context—whether Prisma handles it or the database—can lead to subtle discrepancies in edge cases, especially in distributed systems where time-sensitive operations may be affected by network latency or differing system clocks.

**Reply by kiwavi** [reply]
> Hi @kiwavi
> 
> When you use `@default(now())` in Prisma, the timestamp is generated by Prisma’s Query Engine at the time the query is executed. This means that the current timestamp is assigned by the Prisma layer, and while it's reflected in the Prisma schema, it’s not directly tied to the database's internal timestamping mechanisms. See this [open issue](https://github.com/prisma/prisma/issues/1831)
> 
> In contrast, `@default(dbgenerated("NOW()"))` relies on the database’s native function (`NOW()` in PostgreSQL) to generate the timestamp. In this case, the database itself is responsible for assigning the current timestamp when the record is created, ensuring the timestamp is generated at the database level.
> 
> Although `NOW()` and `CURRENT_TIMESTAMP` are functionally equivalent in PostgreSQL (both return the current timestamp), the key difference lies in how and where they are applied:
> 
> `@default(now())`: Timestamp is assigned by Prisma’s Query Engine, which might introduce minor differences due to application-server clock variations or latency. `@default(dbgenerated("NOW()"))`: Timestamp is assigned natively by the database, ensuring consistency for records created directly within the database, even outside Prisma.
> 
> This difference in context—whether Prisma handles it or the database—can lead to subtle discrepancies in edge cases, especially in distributed systems where time-sensitive operations may be affected by network latency or differing system clocks.

Hi @ludralph, thanks for the answer. I need a clarification though. So what matters is what's defined at the prisma schema (default(now) vs @default(dbgenerated("NOW()"))) and not what they map to in postgres (TIMESTAMP(3) NOT NULL DEFAULT NOW() VS TIMESTAMP(3) NOT NULL DEFAULT CURRENT_TIMESTAMP)

**Reply by ludralph** [reply]
Yes, what matters is how the defaults are defined in the Prisma schema, as this determines where and how the timestamp is generated. `@default(now())` is handled by Prisma itself. The timestamp is generated by the Query Engine at the time of the operation. `default(dbgenerated("NOW()"))` is handled directly by the database. The timestamp is generated by PostgreSQL itself when the record is inserted. You can also take a look at this closed [issue](https://github.com/prisma/prisma/issues/12572) which discusses similar concerns about timestamp differences.
